compiler/backend/src/org/jetbrains/kotlin/codegen/coroutines/coroutines-codegen.md

```diff
@@ -0,0 +1,3263 @@
+when a wrapped coroutine suspends, `getOrThrow` tells `suspendCoroutine` to suspend.
+
+### startCoroutine
+We have already covered how a coroutine suspends, what happens when it resumes and how the compiler handles it. However, we have never
+looked at how one can create or start a coroutine. In all previous examples, one could notice a call to `startCoroutine`. There are two
+versions of the function: to start a suspend lambda without parameters and to start a coroutine with either one parameter or a receiver. It
+is defined as follows:
+```kotlin
+public fun <T> (suspend () -> T).startCoroutine(completion: Continuation<T>) {
+    createCoroutineUnintercepted(completion).intercepted().resume(Unit)
+}
+```
+So, it
+1. creates a coroutine
+2. intercepts it
+3. starts it
+
+Once again, `createCoroutineUnintercepted` has two versions - without parameters and with exactly one parameter. All it does is calling
+suspending lambda's `create` function. After the interception, we resume the coroutine with a dummy value. As explained in the resume with
+the value section, the state-machine ignores its first state value. Thus, it is the perfect way to start a coroutine without calling
+`invokeSuspend`. However, the way we start callable references is different. Since they are tail-call, in other words, do not have
+a continuation inside an object, we wrap them in a hand-written one.
+
+#### create
+`create` is generated by the compiler, and it
+1. creates a copy of the lambda by calling a constructor with captured variables
+2. puts `create`'s arguments into parameter fields.
+
+For example, if we have a lambda like
+```kotlin
+fun main() {
+    val i = 1
+    val lambda: suspend (Int) -> Int = { i + it }
+}
+```
+the generated `create` will look like
+```kotlin
+public fun create(value: Any?, completion: Continuation): Continuation {
+    val resutl = main$lambda$1(this.$i, completion)
+    result.I$0 = value as Int
+}
+```
+note that the constructor, in addition to captured parameters, accepts a completion object.
+
+In Old JVM BE, `create` is generated for every suspend lambda even when we do not need it. I.e., even for suspending lambdas with
+more than one parameter. There are only two versions of `createCoroutineUnintercepted`, and there are no other places where we call
+`create` (apart from compiler-generated `invoke`s). Thus, in JVM_IR BE, we fixed the slip-up, and it generates the `create` function only
+for functions with zero on one parameter.
+
+##### Lambda Parameters
+We need to put the suspend lambda arguments into fields since there can be only one argument of `invokeSuspend` - `$result`. The compiler
+moves the lambda body into `invokeSuspend`. Thus, `invokeSuspend` does all the computation. We reuse fields for spilled variables for
+parameters as well. For example, if we have a lambda with type `suspend Int.(Long, Any) -> Unit`, then `I$0` hold value of extension
+receiver,' `J$0` - the first argument, `L$1` - the second one.
```

Comment:
> receiver,'
The last quote looks dangling, i.e., not matched with any other quote mark around here.

Reply:
> receiver,'

The last quote looks dangling, i.e., not matched with any other quote mark around here.